src/chains/ethereum/ethereum/src/helpers/simulation-handler.ts

```diff
@@ -0,0 +1,479 @@
+import Common from "@ethereumjs/common";
+import VM from "@ethereumjs/vm";
+import { InterpreterStep } from "@ethereumjs/vm/dist/evm/interpreter";
+import { DefaultStateManager } from "@ethereumjs/vm/dist/state/index";
+import { Address } from "@ganache/ethereum-address";
+import { calculateIntrinsicGas } from "@ganache/ethereum-transaction";
+import { CallError } from "@ganache/ethereum-utils";
+import { BUFFER_EMPTY, Data, hasOwn, keccak, Quantity } from "@ganache/utils";
+import {
+  makeStepEvent,
+  VmAfterTransactionEvent,
+  VmBeforeTransactionEvent,
+  VmConsoleLogEvent,
+  VmStepEvent
+} from "../provider-events";
+import {
+  Address as EthereumJsAddress,
+  BN,
+  KECCAK256_NULL
+} from "ethereumjs-util";
+import { GanacheTrie } from "./trie";
+import Blockchain from "../blockchain";
+import { Block, RuntimeBlock } from "@ganache/ethereum-block";
+import { EVMResult } from "@ethereumjs/vm/dist/evm/evm";
+import {
+  AccessList,
+  AccessLists
+} from "@ganache/ethereum-transaction/src/access-lists";
+import { ERROR, VmError } from "@ethereumjs/vm/dist/exceptions";
+import { warmPrecompiles } from "./precompiles";
+import { maybeGetLogs } from "@ganache/console.log";
+import Emittery from "emittery";
+
+export type SimulationTransaction = {
+  /**
+   * The address the transaction is sent from.
+   */
+  from: Address;
+  /**
+   * The address the transaction is directed to.
+   */
+  to?: Address;
+  /**
+   * Integer of the gas provided for the transaction execution. eth_call consumes zero gas, but this parameter may be needed by some executions.
+   */
+  gas: Quantity;
+  /**
+   * Integer of the gasPrice used for each paid gas
+   */
+  gasPrice: Quantity;
+  /**
+   * Integer of the value sent with this transaction
+   */
+  value?: Quantity;
+  /**
+   * Hash of the method signature and encoded parameters. For details see Ethereum Contract ABI in the Solidity documentation
+   */
+  data?: Data;
+  block: RuntimeBlock;
+  /**
+   * Array of addresses and storage keys.
+   */
+  accessList?: AccessList;
+};
+
+type CallOverride =
+  | Partial<{
+      code: string;
+      nonce: string;
+      balance: string;
+      state: { [slot: string]: string };
+      stateDiff: never;
+    }>
+  | Partial<{
+      code: string;
+      nonce: string;
+      balance: string;
+      state: never;
+      stateDiff: { [slot: string]: string };
+    }>;
+
+export type CallOverrides = {
+  [address: string]: CallOverride;
+};
+
+/**
+ * Stripped down version of the type from EthereumJs
+ */
+interface RunCallOpts {
+  caller?: EthereumJsAddress;
+  data?: Buffer;
+  gasPrice?: BN;
+  gasLimit?: BN;
+  to?: EthereumJsAddress;
+  value?: BN;
+  block?: any;
+}
+
+export default class SimulationHandler extends Emittery<{
```

Comment:
The interface for this is odd to me. As it maintains internal state in some conflicting ways

For example:

```javascript
const sim = new SimulationHandler( ... );
sim.on("ganache:vm:tx:before", event => {
   this.emit("ganache:vm:tx:before", event);
});
sim.on("ganache:vm:tx:after", event => {
   this.emit("ganache:vm:tx:after", event);
});
await sim.initialize(this, ... ); // can only be used if we have access to a Blockchain
await sim.runCall(); // emits events with a unique `context`

// emits events for a `context` that's already been tied do another transaction lifecycle!
await sim.runCall();

// we can get really weird and racy if we think about how this interface could be used:
await Promise.all([
  sim.createAccessList(...),
  sim.createAccessList(...)
]);
```

Reply:
I've changed the simulation handler to:
 1. Only allow initialization once (it just logs and returns if you try to run it a second time)
 2. Each `runCall` emits the before/step/after events and makes a new context during the before event
Do you think this is sufficient in fixing what you're getting at?

Reply:
It still seems a bit strange that calling `sim.runCall()` twice may produce different results (since runCall has side effects).

I think this functionality rolled up into a class isn't neccessary, as the state it holds is ephemeral and the class itself is never passed around or used outside of it's own scope.

It's always used in this pattern:

```javascript
const sim = new SimulationHandler(...);
await sim.initialize(...);
const result = await sim.runCall();
```

Whereas, if you exported functions instead, e.g.,

```
import * as Simulations from "./simulation-helpers"

// ...

const result = await Simulations.runCall(options});
const accessList = await Simulations.createAccessList(options});
```

where internally `runCall` and `createAccessList` were something like:

```typescript

// some horrific example types:
type SimulatorVmOptions = { ... }; // i'm very bad at naming things
type CommonOptions = SimulatorVmOptions & { ... };
export type RunCallOptions = CommonOptions & { ... };
export type AccessListOptions = CommonOptions & { ... };
type GetSimulatorVmOptions = { ... };


const createCallSimulatonVm = async (options: SimulatorVmOptions) => {
  // ...
};
export const runCall = async (options: RunCallOptions) => {
   const vm = await createCallSimulatonVm(options);
   // do the things that only runCall needs to do
}
export const createAccessList = async (options: AccessListOptions) => {
  const vm = await createCallSimulatonVm(options);
  // do the things that only createAccessList needs to do
}
```

then we, your dearest readers, won't have to worry about the life of the class.